### PR TITLE
Real NVP Transformation Layer for Normalizing Flows

### PR DESCRIPTION
**Pull request by JoseAntonioSiguenza**
## Description

Following implementing a Normalizing Flow model using Pytorch, this PR includes a new layer/node known as [Real NVP.](http://www0.cs.ucl.ac.uk/staff/A.Sztrajman/webpage/blog/rnvp/rnvp.html) The main advantage of this transformation layer is the ease of computing the inverse pass. It is important to mention that the scope of this PR is to include the layer with its model, which is slightly different from the existing `NormalizingFlow` model.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [x ] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/layers.py [hunk 1]
```diff
@@ -1021,3 +1026,108 @@ def forward(
...
+
+    This particular transformation is represented by the following function:
+    y = x + (1 - x) * exp( s(x)) + t(x), where t and s needs an activation
+    function. This class also returns the logarithm of the jacobians
+    determinant which is useful when invert a transformation and compute
+    the probability of the transformation.
+
+    Parameters
+    ----------
+    input shape: (samples, dim)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:1099]: This should match the numpydoc style. Take a look at

```
Returns
--------
output shape of ...
```

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -1021,3 +1026,108 @@ def forward(
...
+  def inverse(self, y: Sequence) -> Tuple[torch.Tensor, torch.Tensor]:
+    """ Inverse pass
+
+    This class performs the inverse of the previous method (formward).
+    Also, this metehod returns the logarithm of the jacobians determinant
+    which is useful to compute the learneable features of target distribution.
+
+    Parameters
+    ----------
+    input shape: (samples, dim)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:1121]: Same comment here

Comment (JoseAntonioSiguenza) [reply deepchem/models/torch_models/layers.py]: The changes have been added ✅

File: deepchem/models/torch_models/__init__.py [hunk 3]
```diff
@@ -9,5 +9,7 @@
 from deepchem.models.torch_models.pagtn import Pagtn, PagtnModel
 from deepchem.models.torch_models.mat import MAT, MATModel
 from deepchem.models.torch_models.megnet import MEGNetModel
+from deepchem.models.torch_models.normalizing_flows_pytorch import NormalizingFlow
+from deepchem.models.torch_models.layers import CNNModule, CombineMeanStd, WeightedLinearCombo, AtomicConvolution
```
Comment (rbharath) [inline deepchem/models/torch_models/__init__.py:13]: Looks like a duplicated line?

File: deepchem/models/torch_models/layers.py [hunk 4]
```diff
@@ -1189,8 +1188,11 @@ def forward(self, x: Sequence) -> Tuple[torch.Tensor, torch.Tensor]:
     which is useful when invert a transformation and compute the
     probability of the transformation.
 
+    Parameters
+    ----------
     input shape: (samples, dim)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:1193]: Formatting here is a bit off:

```
x: Sequence
 Should have input shape (samples, dim)
```

File: deepchem/models/torch_models/layers.py [hunk 5]
```diff
@@ -1208,8 +1210,11 @@ def inverse(self, y: Sequence) -> Tuple[torch.Tensor, torch.Tensor]:
     useful when invert a transformation and compute the probability of
     the transformation.
 
+    Parameters
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:1213]: Same here

File: deepchem/models/torch_models/layers.py [hunk 6]
```diff
@@ -1664,6 +1669,129 @@ def forward(
...
+
+  Example
+  -------
+  >>> import torch
+  >>> import torch.nn as nn
+  >>> from deepchem.models.torch_models.layers import RealNVPLayer
+  >>> layers = 4
+  >>> hidden_size = 16
+  >>> mask = nn.functional.one_hot(torch.tensor([i % 2 for i in range(layers)])).float()
+  >>> rnvp_model = RealNVPLayer(mask, hidden_size)
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:1689]: Can you add an example of calling the layer?

File: docs/source/api_reference/models.rst [hunk 7]
```diff
@@ -403,6 +403,11 @@ MATModel
 .. autoclass:: deepchem.models.torch_models.MATModel
   :members:
 
+NormalizingFlowModel
+--------------------
+
+.. autoclass:: deepchem.models.torch_models.NormalizingFlow
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:409]: Can you add the new layer to the docs as well?

Comment (JoseAntonioSiguenza) [reply docs/source/api_reference/models.rst:409]: I think this new layer is already added at `docs/source/api_reference/layers.rst`

**Comment by JoseAntonioSiguenza** [comment]
Hi @rbharath @paupaiz! I've checked the failing CI tests and I think they are not related to this PR. Can you help me with a review? 🙌🏼
